cam6_3_127:  Turn on HB where CLUBB is not active

jtruesdal opened the pull request:
Free atmosphere Richardson number based mixing (using the free atmosphere part of the Holtslag-Boville PBL scheme) for those layers where CLUBB is not active (turned on in cam6 or cam_dev). Also helps stabilize less diffusive dynamical cores. A new namelist variable will be added to allow this new functionality.

The issue for the PR is #846.  This PR will also address issue #772

The science validation for this code is:
https://webext.cgd.ucar.edu/FLTHIST/f.cam6_3_112.FLTHIST_v0c.ne30.non-ogw-ubcT-effgw0.3-rdg_beta0.5-vtrmi1.5-hb-tt.001/atm/

closes #772
closes #846

jtruesdal commented:
I am making some minor corrections due to regression test failures but will "undraft" this PR very soon.

jtruesdal commented:
@nusbaume Ready for initial review.  This PR is only for using HB scheme with cam_dev.

nusbaume commented on src/physics/cam/hb_diff.F90:
```diff
@@ -247,6 +248,96 @@ subroutine compute_hb_diff(lchnk, ncol,            &
+    real(r8), intent(in)  :: shflx(pcols)             ! sensible heat flux
+    real(r8), intent(in)  :: qflx(pcols)              ! water vapor flux
+    real(r8), intent(in)  :: pmid(pcols,pver)         ! midpoint pressures
+    character(len=16), intent(in) :: eddy_scheme
```
I don't think `eddy_scheme` is being used in this subroutine.

jtruesdal replied:
Removed it here and in the call to compute_hb_free_atm_diff in vertical_diffusion.F90

nusbaume commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -1097,7 +1139,7 @@ subroutine vertical_diffusion_tend( &
      shflux = 0._r8
      cflux(:,1) = 0._r8
-     if (cam_physpkg_is("cam_dev")) then
+     if (clubb_do_hb_above) then
```
Do we actually want to do this?  I thought this change was originally because of the physics re-order in `cam_dev`, and so it might cause problems if someone turns this option on for, say, CAM6 physics.

jtruesdal replied:
@PeterHjortLauritzen Jesse's catch here was actually a bug I introduced when getting the mods ready for the PR. Since it will affect the tests and could be the reason for the cam6 waccm failure I am rerunning the regression tests.  If everything passes we will be able to include CAM6 as part of this PR and be done with it. Thanks again for your careful review Jesse!

PeterHjortLauritzen replied:
My re-creation of the failing regression test (FX2000; CAM_CONFIG_OPTS: -phys cam6 -waccmx -ionosphere wxie -chem waccm_ma_mam5) is still failing with this code change ...

nusbaume reviewed (changes requested):
Looks pretty good!  I only have some (hopefully) minor change requests and a few questions.

nusbaume commented:
Also forgot to ask if this PR will fully resolve issues #772 and #846?  If so it would be good to add the "closes #XXX" keyword to the main PR description box so that those issues are closed automatically once this PR is merged in.

PeterHjortLauritzen commented on src/physics/cam/hb_diff.F90:
```diff
@@ -247,6 +248,96 @@ subroutine compute_hb_diff(lchnk, ncol,            &
+    !
+    ! Input arguments
+    !
+    integer, intent(in) :: lchnk                      ! chunk index (for debug only)
```
Removed (Jesse = NAG compiler :-) )

jtruesdal replied:
removed lchnk in called and calling routines.

PeterHjortLauritzen commented on src/physics/cam/hb_diff.F90:
```diff
@@ -247,6 +248,96 @@ subroutine compute_hb_diff(lchnk, ncol,            &
+    !
+    !---------------------------Local workspace-----------------------------
+    !
+    real(r8) :: thv(pcols,pver)         ! virtual temperature
```
agree. Comment fixed

jtruesdal replied:
Looks to be OK now

PeterHjortLauritzen commented on src/physics/cam/hb_diff.F90:
```diff
@@ -247,6 +248,96 @@ subroutine compute_hb_diff(lchnk, ncol,            &
+    real(r8) :: s2(pcols,pver)          ! shear squared
+    real(r8) :: n2(pcols,pver)          ! brunt vaisaila frequency
+
+    ! virtual temperature
```
fixed

nusbaume replied:
It doesn't look like this change was actually committed/pushed?

PeterHjortLauritzen replied:
Got very distracted and did not push ...

jtruesdal replied:
Fixed.

PeterHjortLauritzen commented on src/physics/cam/hb_diff.F90:
```diff
@@ -247,6 +248,96 @@ subroutine compute_hb_diff(lchnk, ncol,            &
+    kvh(:ncol,:) = kvf(:ncol,:)
+    cgh(:ncol,:) = 0._r8
+    cgs(:ncol,:) = 0._r8
+    return
```
removed (also removed from subroutine compute_hb_diff on which this subroutine is based)

nusbaume replied:
It doesn't look like this change was actually committed/pushed?

jtruesdal replied:
Looks like its OK now

PeterHjortLauritzen commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -276,7 +276,8 @@ subroutine vertical_diffusion_init(pbuf2d)
   use beljaars_drag_cam, only : beljaars_drag_init
   use upper_bc,          only : ubc_init
-  use phys_control,      only : waccmx_is, fv_am_correction
+  use phys_control,      only : waccmx_is, fv_am_correction, cam_physpkg_is
```
correct. removed.

PeterHjortLauritzen commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -401,6 +402,19 @@ subroutine vertical_diffusion_init(pbuf2d)
      call addfld('HB_ri',      (/ 'lev' /),  'A', 'no',  'Richardson Number (HB Scheme), I' )
   case ( 'CLUBB_SGS' )
      do_pbl_diags = .true.
+       call init_hb_diff(gravit, cpair, ntop_eddy, nbot_eddy, pref_mid, &
```
fixed

jtruesdal replied:
Looks to be fixed now.

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,4 +1,130 @@
+Describe any changes made to build system: NA
+
+Describe any changes made to the namelist:
+  - Introduced namelist variable clubb_do_hb_above. Default False unless using cam6 or cam_dev then
```
Yes.  Corrected the comment.  We will add cam6 when waccmx is working with this scheme.

jtruesdal replied:
As the cam6 fix is now included for using HB above CLUBB I changed the comment to reflect that it can be used for cam_dev and cam6.

jtruesdal commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -401,6 +402,19 @@ subroutine vertical_diffusion_init(pbuf2d)
+       call init_hb_diff(gravit, cpair, ntop_eddy, nbot_eddy, pref_mid, &
+            karman, eddy_scheme)
+     !
+     ! run HB scheme where CLUBB is not active when running cam6 or cam_dev
```
Yep. Corrected the comment.

jtruesdal replied:
As the cam6 fix is now included for using HB above CLUBB I changed the comment to reflect that it can be used for cam_dev and cam6.

jtruesdal commented:
I modified the description to note that this PR currently applies to cam_dev and the issues noted will be kept open until a future PR addresses a couple cam6 regression errors.  The issues will be fully resolved and closed in that future PR.

PeterHjortLauritzen commented:
> I think some changes were either not committed or not pushed. I tried to mark which changes were still "missing", but if it is unclear which changes those are just let me know. Thanks!

Sorry had a several day delay in pushing changes ...

jtruesdal commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -4296,6 +4296,13 @@ Real: number of plumes in mass flux ensemble
 
+<entry id="clubb_do_hb_above" type="logical" category="pbl"
+       group="clubb_params_nl" valid_values="" >
+Logical: If True activate Holtslag and Boville PBL scheme where CLUBB is not active
```
Replaced PBL with "vertical diffusion"

nusbaume reviewed (approved):
Everything looks good to me now, thanks!

cacraigucar commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -4296,6 +4296,13 @@ Real: number of plumes in mass flux ensemble
 Default: 10
 </entry>
 
+<entry id="clubb_do_hb_above" type="logical" category="pbl"
```
I propose changing the name of this to be `do_hb_above_clubb`.   There are several reasons I believe this is advantageous:

1. It reads more like what it is doing (initially I saw this flag indicating clubb doing somthing in regards to hb)
2. This variable is really controlling hb and not clubb
3. All other `clubb_` namelists are controlling clubb behaviour

jtruesdal replied:
Changed clubb_do_hb_above to do_hb_above_clubb and moved definition from clubb control section of namelist to "PBL and vertical diffusion" section.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -805,8 +811,9 @@ subroutine clubb_readnl(nlfile)
-         clubb_wpxp_L_thresh
-                               
+         clubb_wpxp_L_thresh, &
+         clubb_do_hb_above
```
I believe on my argument above, that this is actually controlling hb behaviour and not clubb.  If that is indeed the case, then I propose this namelist should be in the hb_diff module (and a namelist reader will need to be added to that module).

jtruesdal replied:
Sounds good to me. if @PeterHjortLauritzen agrees I'll make the changes.

brian-eaton replied:
I would suggest putting this in phys_ctl_nl.  That's a good place for a control variable that's needed in both clubb_intr and vertical_diffusion.  Putting it in vertical_diffusion is also an option since that's where the control layer is for vertical diffusion.  But clubb_intr can't access it directly from vertical diffusion because of circular dependency issues.  Another option is to let vertical_diffusion_register add clubbtop to the pbuf, then have clubb_ini_cam check to see whether the index is in the pbuf.  But I think putting it in phys_ctl_nl is probably the easiest, and most readily understandable.

jtruesdal replied:
Moved HB control variable to to phys_ctl_nl. Now variable accessed via phy_getopts call like other physics control variables.

brian-eaton replied:
This is OK, but I'd like to point out that more recently we've just been accessing variables from the `phys_control` module by giving them `public, protected` attributes.  It's easier than adding arguments to the accessor methods which were implemented before the `protected` attribute was available.

jtruesdal replied:
I didn't notice that we had some public protected in there which is cleaner and easier.  I will update to that method as phys_control is already a mix of the two.  A little less to cleanup in the future.

cacraigucar reviewed (changes requested):
Need to use a keyword in the description to automatically close the issues.  `closes #XXX` is an example.

Submitting these comments early as they may take a bit to implement.  I plan on doing a more complete review, so more comments may be forthcoming (I also want to get feedback on whether my understanding is correct or not).

cacraigucar reviewed (changes requested):
Next round of change requests

jtruesdal commented on src/dynamics/se/dycore/global_norms_mod.F90:
```diff
@@ -681,7 +681,9 @@ subroutine print_cfl(elem,hybrid,nets,nete,dtnu,ptop,pmid,&
       ! Vertical profile from FV dycore (see Lauritzen et al. 2012 DOI:10.1177/1094342011410088)
       scale1        = 0.5_r8*(1.0_r8+tanh(2.0_r8*log(pmid(sponge_del4_lev)/pmid(k))))
-      nu_div_lev(k) = (1.0_r8-scale1)*nu_div+scale1*nu_div_max
+      if (sponge_del4_nu_div_fac.ne.1.0_r8) then
```
Fixed both instances.

jtruesdal commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -979,42 +995,67 @@ subroutine vertical_diffusion_tend( &
+    ! run HB scheme where CLUBB is not active when running cam_dev
+    !
+    if (clubb_do_hb_above) then
+      call compute_hb_free_atm_diff( lchnk     , ncol          , &
```
Jesse noted that lchnk was not used in compute_hb_diff and hb_free_atm_diff and I removed them but apparently missed this parameter in the calling routine.  It's fixed now.

brian-eaton commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -979,42 +995,67 @@ subroutine vertical_diffusion_tend( &
+           cam_in%shf, cam_in%cflx(:,1), obklen  , ustar       , &
+           kvm       , kvh      , kvq     , cgh      , cgs     , &
+           ri        )
+      clubbtop_idx = pbuf_get_index('clubbtop')
```
The call to pbuf_get_index belongs in vertical_diffusion_init (so it's only done once).

jtruesdal replied:
Moved call to pbuf_get_index as recommended.

brian-eaton commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -649,6 +661,7 @@ subroutine vertical_diffusion_tend( &
   ! This is an interface routine for vertical diffusion !
   !---------------------------------------------------- !
   use physics_buffer,     only : physics_buffer_desc, pbuf_get_field, pbuf_set_field
+  use physics_buffer,     only : pbuf_get_index
```
This should be removed along with the corresponding call below (belongs in init method).

jtruesdal replied:
Updated as recommended.

brian-eaton commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -839,6 +853,8 @@ subroutine vertical_diffusion_tend( &
   real(r8) :: tauy(pcols)
   real(r8) :: shflux(pcols)
   real(r8) :: cflux(pcols,pcnst)
+  integer  :: clubbtop_idx
```
This should be private module data (initialized by init method).

jtruesdal replied:
Moved to private data.

jtruesdal commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -467,8 +468,10 @@ subroutine clubb_register_cam( )
+    use physics_buffer,  only: pbuf_add_field, dtype_r8, dtype_i4, dyn_time_lvls, pbuf_get_index
     use subcol_utils,    only: subcol_get_scheme
+
+    integer :: err_idx, idx
```
Removed.

jtruesdal commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -467,8 +468,10 @@ subroutine clubb_register_cam( )
 
     !  Add CLUBB fields to pbuf 
-    use physics_buffer,  only: pbuf_add_field, dtype_r8, dyn_time_lvls
+    use physics_buffer,  only: pbuf_add_field, dtype_r8, dtype_i4, dyn_time_lvls, pbuf_get_index
```
Removed.

jtruesdal commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -205,7 +204,7 @@ subroutine vd_register()
   !------------------------------------------------ !
 
-  use physics_buffer,      only : pbuf_add_field, dtype_r8, dtype_i4
+  use physics_buffer,      only : pbuf_add_field, dtype_r8, dtype_i4, pbuf_get_index
```
Removed.

jtruesdal commented:
@cacraigucar @nusbaume @brian-eaton Just a heads up: there is new code to look at for your rereview in addition to mods related to your comments.  Peter corrected the bug which caused cam6 FV regression failures.  The new mods are in geopotential.F90.  Previously do_hb_above_clubb was only enabled for cam_dev but now that we've corrected the error I am also enabling do_hb_above_clubb for cam6 as well.  Running HB above clubb is now the default for all dycores running cam6 or cam_dev physics.  Jesse, I unresolved 2 comments of yours that referenced incorrect comments that were modified with this latest addition (I added cam6 to the comment along with cam_dev.)

nusbaume reviewed (approved):
The new changes look good to me!

fvitt commented on doc/ChangeLog:
```diff
@@ -1,4 +1,144 @@
+appropriate machine below.  All failed tests must be justified.
+
+cheyenne/intel/aux_cam: all BFB except:
+  ERP_Ln9_Vnuopc.ne30pg3_ne30pg3_mg17.FW2000climo.cheyenne_intel.cam-outfrq9s_wcm_ne30 (Overall: FAIL)
+     - pre-existing failure
```
Is HB active above clubb in this test case?  Shouldn't HB mixing stabilize WACCM-SE?  Does this need a new spunup IC file?

jtruesdal replied:
@PeterHjortLauritzen Do you have any insight into this failure? I think it is a crossing of levels failure. The HB mixing will stabilize WACCM-SE but it may not be enough to overcome an out of balance IC file. (If that is the issue)

PeterHjortLauritzen replied:
@fvitt Thanks for the comment/reminder: 

Yes HB helps stabilize the initial condition but the existing initial condition uses an older topography. With the new topography (where PHIS=0 over ocean) some spin-up with more damping is necessary (due to the gravity waves excited by the change in topography). Here is a spun-up initial condition:

/glade/p/cgd/amp/pel/inic/FW2000.ne30pg3_ne30pg3_nlev70.i.nc

(I did not know that when running with full chemistry the model does not run when using an FWsc2000 initial conditions - hangs at startup with plenty error messages in the CESM log file; so I took the default IC file and overwrote U,V,T,Q, PS with the spun-up fields from an FWsc2000 IC file I had: ncks -m -A -C -v U,V,T,Q,PS spinup.ic.nc orig.ic.nc)

For the failing test:

SMS_Ln9_Vnuopc.ne30pg3_ne30pg3_mg17.FW2000climo.cheyenne_intel.cam-outfrq9s_wcm_ne30

we can now use the default se_nsplit. (no need to increase to 30 anymore).

Need some guidance to make these changes ... or maybe @jtruesdal can do it.

fvitt replied:
@PeterHjortLauritzen 
Full TSMLT chemistry can be tricky to spin up if all the chemical constituents are starting from zero concentrations (no chemical tracers in the IC file). The chemical solver may not converge in some places.  Are these the type of error messages you are seeing?
I am not sure what you need guidance on.  It sounds like you have a working IC by using the dynamical fields taken from a working FWsc2000 IC file...

PeterHjortLauritzen replied:
Yes exactly those errors ...

I have everything I need ... but I don't think I have permissions to add new IC files etc. I think @jtruesdal can take care of it ...

PeterHjortLauritzen replied:
... just realized I did not add the mandatory meta data to IC file and change the file name to incl. creation date!

Modified IC file now:

/glade/p/cgd/amp/pel/inic/FW2000.ne30pg3_ne30pg3_nlev70_c230906.nc

jtruesdal replied:
I will check over the new IC file and load it onto our local systems.  I'll submit it to the data repository once the PR is approved.

jtruesdal replied:
SMS and ERP WACCM tests now passing using the new IC and default se_nsplit

nusbaume commented on src/dynamics/se/dycore/prim_advance_mod.F90:
```diff
@@ -1176,51 +1176,57 @@ subroutine compute_and_apply_rhs(np1,nm1,n0,dt2,elem,hvcoord,hybrid,&
+
+            if (hvcoord%hybm(k)>0) then
+               call gradient_sphere(log(exner(:,:)),deriv,elem(ie)%Dinv,grad_logexner)
+               pgf_term(:,:,1)=pgf_term(:,:,1) + &
+                    cpair*T0*(grad_logexner(:,:,1)-pgf_term(:,:,1)/exner(:,:))
+               pgf_term(:,:,2)=pgf_term(:,:,2) + &
+                    cpair*T0*(grad_logexner(:,:,2)-pgf_term(:,:,2)/exner(:,:))
```
Is this formulation still equivalent to the previous one?  It looks like `grad_exner` variable was still just the gradient of the exner function at this point, but the `pgf_term` variable now includes a `cp_dry(:,:,k)*theta_v(:,:)` term as well.

jtruesdal replied:
I agree. @PeterHjortLauritzen could you double check the implementation for these lines?

PeterHjortLauritzen replied:
Thanks @nusbaume ! Indeed a bug - I have verified that answers are bit-for-bit with the code version before the buggy commit.

nusbaume replied:
Thanks @jtruesdal and @PeterHjortLauritzen!  The new code looks good to me.

nusbaume reviewed (changes requested):
Looks good, but I did have one question that would be great for someone to sanity-check me on.

fvitt commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -5116,6 +5116,13 @@ Bretherton; 'HBR' for Rasch modified version of 'HB'.
+       group="phys_ctl_nl" valid_values="" >
+Logical: If True activate Holtslag and Boville vertical diffusion scheme where CLUBB is not active
+         (note that CLUBB top is dynamic in each column)
+Default: none
```
Should the default be "FALSE" or "set by build-namelist"?

PeterHjortLauritzen replied:
Default in namelist_defaults_cam.xml is false. Please check if I modified the code correctly ... thanks

brian-eaton replied:
It is done correctly, but the default depends on the physics package.  So the documentation in `namelist_definition.xml` should read:
`Default: Set by build-namelist.` instead of `Default: none`

PeterHjortLauritzen replied:
Thanks. Changed to "Set by build-namelist."

fvitt commented on src/dynamics/se/dycore/control_mod.F90:
```diff
@@ -124,5 +124,7 @@ module control_mod
-  
+
+
+  integer, public :: pgf_formulation !PGF formulation - see prim_advance_mod.F90
```
Initialize this to -1 here to ensure that it get properly set later
```suggestion
  integer, public :: pgf_formulation = -1 !PGF formulation - see prim_advance_mod.F90
```

fvitt commented on src/dynamics/se/dycore/prim_advance_mod.F90:
```diff
@@ -1177,51 +1176,50 @@ subroutine compute_and_apply_rhs(np1,nm1,n0,dt2,elem,hvcoord,hybrid,&
-         end if
-
-
+           pgf_term(:,:,1)  = density_inv(:,:)*grad_p_full(:,:,1,k)
+           pgf_term(:,:,2)  = density_inv(:,:)*grad_p_full(:,:,2,k)
```
Is this block for pgf_formulation==2? Should check for pgf_formulation==2.  Error out if pgf_formulation is not 1 or 2.

PeterHjortLauritzen replied:
Done!

fvitt commented on src/physics/cam/vertical_diffusion.F90:
```diff
@@ -599,6 +616,9 @@ subroutine vertical_diffusion_init(pbuf2d)
      kvh_idx = pbuf_get_index('kvh')
   end if
 
+  ! pbuf field denoting top of clubb
+  clubbtop_idx = pbuf_get_index('clubbtop')
```
Is the `clubbtop` pbuf field always available?
```suggestion
  if (do_hb_above_clubb) then
     ! pbuf field denoting top of clubb
     clubbtop_idx = pbuf_get_index('clubbtop')
  end if
```

PeterHjortLauritzen replied:
I think it is always available when using CLUBB

brian-eaton replied:
`clubb` is only on by default for `cam6/cam_dev`.  So I think the original code will fail for pre-cam6 configurations.  Francis's suggested change is one approach.  Another would be to use the optional `errcode` argument to `pbuf_get_index` so that if it's not available the function will just return -1 rather than `call endrun`.

fvitt reviewed (changes requested):
A few suggestions
